src/psbt.js

```diff
@@ -231,6 +234,135 @@ export function psbtOutputFormatter(output) {
+  // To provide the non-PSBT wallet with the proper info, we actually
+  // need to grab some info from the "data inputs" and some other
+  // information from the "transaction inputs".
+  let txInputs = localPSBT.txInputs.map((input, index) => {
+    const dataInput = localPSBT.data.inputs[index];
+    // The bip32Derivation hanging off of the dataInput contains
+    // an array of objects like:
+    //    [
+    //      { masterFingerprint1, pubkey1, path1 },
+    //      { masterFingerprint2, pubkey2, path2 },
+    //      ...
+    //    ]
+    // We need to use the information in the keyDetails object
+    // to filter down to the fingerprint+pubkey+paths that we care about.
+    // E.g. which key is trying to sign?
+
+    const bip32Derivation = dataInput.bip32Derivation.filter((b32d) => {
+      if (b32d.path.startsWith(keyDetails.root)
+        && b32d.masterFingerprint.toString('hex') === keyDetails.xfp)
+        return b32d
+    });
+
+    if (!bip32Derivation.length) {
+      throw new Error("keyDetails not included in PSBT");
+    }
+
+    bip32Derivations.push(bip32Derivation[0]);
+    // FIXME - this is where we're currently only handling P2SH correctly
+    const fundingTxHex = dataInput.nonWitnessUtxo.toString('hex');
+    const fundingTx = Transaction.fromHex(fundingTxHex);
+    const multisig = generateMultisigFromHex(network, addressType, dataInput.redeemScript.toString('hex'));
+
+    return {
+      amountSats: fundingTx.outs[input.index].value,
+      index: input.index,
+      transactionHex: fundingTxHex,
+      txid: reverseBuffer(input.hash).toString('hex'),
+      multisig,
+    }
+  });
+  let txOutputs = localPSBT.txOutputs.map(output => {
+    return {
+      address: output.address,
+      amountSats: output.value,
+    }
+  });
+
+  return {
+    txInputs,
+    txOutputs,
+    bip32Derivations
+  }
+}
+
+/**
+ * Given a PSBT, an input index, a pubkey, and a signature,
+ * update the input inside the PSBT with a partial signature object.
+ *
+ * Make sure it validates, and then return the PSBT with the partial
+ * signature inside.
+ *
+ * @param {Object} psbt - Psbt Object from bitcoinjs-lib (unsigned or partially signed)
+ * @param {number} inputIndex - which input is this signature for
+ * @param {Buffer} pubkey - public key associated with signature
+ * @param {Buffer} signature - signature of transaction for pubkey
+ * @return {Object} - validated PSBT Object with an added signature for given input
+ * @private
+ */
+function _addSignatureToPSBT(psbt, inputIndex, pubkey, signature) {
+  const partialSig = [
+    {
+      pubkey,
+      signature,
+    }
+  ]
+  psbt.data.updateInput(inputIndex, {partialSig})
+  if (!psbt.validateSignaturesOfInput(inputIndex, pubkey))
+    throw new Error("One or more invalid signatures.");
+  return psbt;
+}
+
+/**
+ * Given an unsigned PSBT, an array of signing public key(s) (one per input),
+ * an array of signature(s) (one per input) in the same order as the pubkey(s),
```

Comment:
Agree it's somewhat brittle. That said -- this isn't really how signing is supposed to work on non-psbt devices. The data used here actually comes from two different places. We actually create the public key array in the constructor of the interaction class because we have the info at that time (in the derivations) - it is created in the order of the inputs as we're looping through them then.

The signatures that come back from the device as an array are also in the order of the inputs - the device itself doesn't send back public keys after signing and we can't really control what it sends back to us.

The memory of order assumption hell comes about where we are doing a sort of inverse of what's happening here -- e.g. a partially- or fully-signed PSBT comes back from a device and instead of just **using** that psbt, we want to pull the signatures out of it to create a pubkey keyed dictionary of signature arrays (... which must be ordered (properly!) to work given the signature validation scheme) to match what the consuming application expected. More refactoring/dropping of that kind of code likely in the future.

Reply:
Thanks for the recall! That makes sense.